### PR TITLE
Bootstrapping webpack repository setup for contributors

### PR DESCRIPTION
**Pull request by Legends**
**What kind of change does this PR introduce?**
Adds a folder setup including two `js` files (`setup.js` & `index.js`) to the project root.
Adds a `setup` entry to package.json `scripts` section and modifies `pretest` entry to incorporate the setup.

What exactly does the setup script (purpose)? --> see section: **Summary**

**Did you add tests for your changes?**
No, difficult to test a setup script using unit-tests. Advise appreciated. 
I guess this will be accomplished through `travis-ci`.

**If relevant, link to documentation update:**
Documentation not yet updated...

**Summary**
[Link to issue/discussion](https://github.com/webpack/webpack/issues/6830)

The developer now just has to clone the webpack repository and hit `yarn test`, in order to execute the tests. 
No manual setup needed anymore:

> ~~cd webpack~~
> // The setup script does not install `yarn` currently! But it can be implemented,
> // if we decide in the future to completely switch to `yarn`
> (npm install -g yarn)  
> ~~yarn install~~
> ~~yarn link~~
> ~~yarn link webpack~~

**Setup.js - Initial draft- How does it work**

The current flow of the setup script is:

CLI: >> `yarn test`

 1. Check if a webpack symlink already exists, if yes --> skip setup..
 2. If no, check if Yarn is installed, if yes: use `yarn` otherwise `npm` (fallback)
 3. Execute `yarn|npm install` and setup `symlink` to webpack
 4. Verify if the `symlink` was established successfully
 5. Show Success/Error Message --> call `process.exit(0|1)`
 6. If everything is ok, tests are being executed

The `setup.js` script is kept very simple without additional dependencies/dev-dependencies.

**Does this PR introduce a breaking change?**
No

**Other information**
Additional functionality can be implemented or existing changed.

### CODE REVIEW

File: setup/setup.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		}).catch((e) => {
+			process.exit(1);
+		});
+		return p;
+	}
+
+	runSetupAsync() {
+		return new Promise((resolve, reject) => {
+			try {		
+				const pm = this.hasYarn ? "yarn" : "npm";
```
Comment (sokra) [inline setup/setup.js]: npm doesn't work, because the lockfile is in yarn format. Only yarn is supported

File: setup/setup.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+					reject(e);
+				});
+			} else {
+				console.log(this.msg.setupSkip);
+				resolve();
+			}
+		});
+
+		p.then(() => {
+			process.exit(0);
```
Comment (sokra) [inline setup/setup.js]: Never use `process.exit`, but use `process.exitCode = 0;` instead.

File: setup/setup.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
+/* eslint-disable */
```
Comment (sokra) [inline setup/setup.js]: That's not good. Remove this and disable rule per line instead if needed.

Also add the `setup` folder to the eslint and prettier scripts.

File: setup/setup.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+"use strict";
+
+class Setup {
```
Comment (sokra) [inline setup/setup.js]: A class is not needed here. Move stuff from constructor into module scope.

File: package.json [hunk 5]
```diff
@@ -87,6 +87,7 @@
     "schemas/"
   ],
   "scripts": {
+    "setup":"node ./setup/index.js",
```
Comment (sokra) [inline package.json]: space between `:` and `"`

File: setup/setup.js (hunk 3 above)
Comment (Legends) [reply setup/setup.js]: removed..

File: setup/setup.js (hunk 4 above)
Comment (Legends) [reply setup/setup.js]: Alright, @montogeek already mentioned this, but still wondering, what's wrong with a class?

File: setup/setup.js (hunk 3 above)
Comment (Legends) [reply setup/setup.js]: Added the setup folder. But `prettier` is hell... not sure if this makes code prettier, and interpreting the error msg is not fun ⏎↹↹↹`  ??? Return with three tabs in the middle of my function name? 
I am using VsCode.. will try it manually..

Comment (Legends) [reply setup/setup.js]: Aha, double click on the error --> Fix all. That's easy.

File: package.json (hunk 5 above)
Comment (Legends) [reply package.json]: done

File: setup/setup.js (hunk 4 above)
Comment (montogeek) [reply setup/setup.js]: Nothing, it is a common practice to put your `require` at the top of the file

**Comment by Legends** [comment]
Git git git, don't know what happend, but I ended up with a new [PR](https://github.com/webpack/webpack/pull/6880) ... sorry.

File: setup/setup.js (hunk 4 above)
Comment (Legends) [reply setup/setup.js]: A that's why, ok. But I still have some `require` statements inside my code, as these requires may or may not be called, depending on the execution flow.

Btw. I have messed something up with Git and have therefore created a new PR [here](https://github.com/webpack/webpack/pull/6880)

**Pull request closed without merging**